kmmsdk/src/commonMain/kotlin/com/lokate/kmmsdk/LokateSDK.kt

```diff
@@ -1,103 +1,227 @@
+    private var appTokenSet: Boolean = false
+
     fun getScanResultFlow(): Flow<BeaconScanResult> {
         return beaconScanner.scanResultFlow()
     }
 
+    fun setAppToken(appToken: String) {
+        lokateScope.launch {
+            authenticationRepository.setAppToken(appToken)
+        }.invokeOnCompletion {
+            log.d { "App token set" }
+            appTokenSet = true
+        }
+    }
+
+    fun fetchActiveBeacons(branchId: String) {
+        val branch =
+            branchId.ifEmpty {
+                log.e { "Branch ID is empty" }
+                "1b224840-de56-41a2-92e0-959193b0035e"
+            }
+        lokateScope.launch {
+            beaconRepository.fetchBeacons(branch).let {
+                if (it is RepositoryResult.Success) {
+                    branchBeacons.addAll(it.body)
+                    log.d { "Branch beacons: $branchBeacons" }
+                }
+            }
+        }
+    }
+
+    private fun checkAppToken() {
+        lokateScope.launch {
+            authenticationRepository.getAppToken().let {
+                if (it is RepositoryResult.Success && it.body.isNotEmpty()) {
+                    appTokenSet = true
+                }
+            }
+        }
+    }
+
+    init {
+        log.d { "LokateSDK initializing" }
+        // check if app token is set
+        checkAppToken()
+
+        if (!appTokenSet) {
+            log.e { "App token not set" }
+            // use default token
+            setAppToken("1")
+        }
+    }
+
     fun startScanning() {
         beaconScanner.setRegions(DEFAULT_BEACONS)
+        //
+        fetchActiveBeacons(EMPTY_STRING)
+        //
         beaconScanner.startScanning()
 
-        val beaconScannerFlow = beaconScanner.scanResultFlow()
-        val newComer = beaconScannerFlow.filter { it !in activeBeacons }
-        val alreadyIn =
-            beaconScannerFlow.filter { it in activeBeacons }.transform {
-                emit(
-                    activeBeacons.find {
-                        it.beaconUUID == it.beaconUUID && it.major == it.major && it.minor == it.minor
-                    }?.copy(lastSeen = getTimeMillis()),
-                )
-            }
-        // in the last 3 seconds, if the beacon is not in the list, it is considered gone
-        val gone =
-            flow {
+        scanResultHandler(
+            beaconScannerFlow =
+                beaconScanner.scanResultFlow().transform { scan ->
+                    val beacon =
+                        branchBeacons.firstOrNull {
+                            it.id.lowercase() == scan.beaconUUID.lowercase() &&
+                                it.major == scan.major.toString() &&
+                                it.minor == scan.minor.toString()
+                        }
+                    val comparison = (beacon == null || beacon.range.ordinal >= scan.proximity.ordinal)
+                    if (scan.accuracy != null && scan.accuracy > -1.0 && comparison) {
```

Comment:
Is it possible for scan.accuracy to be null? I know it is defined as nullable but it is set to the result of calculateAccuracy() function in "kmmsdk/src/androidMain/kotlin/com/lokate/kmmsdk/utils/Beacon.kt", which has the following body:

```
fun calculateAccuracy(
    txPower: Int,
    rssi: Double,
): Double {
    if (rssi == 0.0) {
        return -1.0 // if we cannot determine accuracy, return -1.
    }

    val ratio = rssi * 1.0 / txPower
    return if (ratio < 1.0) {
        ratio.pow(10.0)
    } else {
        val accuracy = 0.89976 * ratio.pow(7.7095) + 0.111
        accuracy
    }
}
```

I suggest we make accuracy field not nullable and discard null checks

Reply:
we may also consider changing its name to distance as in:
http://www.davidgyoungtech.com/2020/05/15/how-far-can-you-go